util/createTrieResolver.js

```diff
@@ -1,108 +1,103 @@
-      const leafTerminatingPaths = basePaths.filter(child => Buffer.isBuffer(child.value))
-      each(leafTerminatingPaths, (child, cb) => {
-        return waterfall([
-          (cb) => leafResolver.util.deserialize(child.value, cb),
-          (ethObj, cb) => leafResolver.resolver._mapFromEthObject(ethObj, options, cb)
-        ], (err, grandChildren) => {
-          if (err) return cb(err)
-          // add prefix to grandchildren
-          grandChildren.forEach((grandChild) => {
-            paths.push({
-              path: child.path + '/' + grandChild.path,
-              value: grandChild.value,
-            })
-          })
-          cb()
-        })
-      }, (err) => {
-        if (err) return callback(err)
-        callback(null, paths)
-      })
-    })
+  if (leafResolver !== undefined) {
+    value = leafResolver.util.deserialize(value)
   }
 
-  // create map from merkle-patricia-tree nodes
-  function mapFromBaseTrie (trieNode, options, callback) {
-    let paths = []
+  return value
+}
 
-    if (trieNode.type === 'leaf') {
-      // leaf nodes resolve to their actual value
-      paths.push({
-        path: nibbleToPath(trieNode.getKey()),
-        value: trieNode.getValue()
-      })
-    }
+// create map from merkle-patricia-tree nodes
+const mapFromBaseTrie = (codec, finalNode, trieNode, leafResolver) => {
+  if (trieNode.type === 'leaf') {
+    const value = getLeafValue(trieNode, leafResolver)
+    addNibbleToObject(finalNode, trieNode.getKey(), value)
+    return
+  }
+
+  trieNode.getChildren().forEach(([nibble, value]) => {
+    let valueToAdd
+    if (EthTrieNode.isRawNode(value)) {
+    // inline child root
+      const childNode = new EthTrieNode(value)
 
-    each(trieNode.getChildren(), (childData, next) => {
-      const key = nibbleToPath(childData[0])
-      const value = childData[1]
-      if (EthTrieNode.isRawNode(value)) {
-        // inline child root
-        const childNode = new EthTrieNode(value)
-        paths.push({
-          path: key,
-          value: childNode
-        })
-        // inline child non-leaf subpaths
-        mapFromBaseTrie(childNode, options, (err, subtree) => {
-          if (err) return next(err)
-          subtree.forEach((path) => {
-            path.path = key + '/' + path.path
-          })
-          paths = paths.concat(subtree)
-          next()
-        })
+      if (childNode.type === 'leaf') {
+        // Make sure the object is nested correctly
+        nibble.push(...childNode.getKey())
+        valueToAdd = getLeafValue(childNode, leafResolver)
       } else {
-        // other nodes link by hash
-        let link = { '/': cidFromHash(multicodec, value).toBaseEncodedString() }
-        paths.push({
-          path: key,
-          value: link
-        })
-        next()
+        valueToAdd = childNode
       }
-    }, (err) => {
-      if (err) return callback(err)
-      callback(null, paths)
-    })
+    } else {
+    // other nodes link by hash
```

Comment:
indent comment by 2

Reply:
That's actually intentionally. I still try to figure out how to format "if/else" comments properly.

My preferred style is:

```
// Some case
if {
}
// And in case of xyz
else {
}
```

But that doesn't work with our linting rules.


```
// Some case
if {
}
else {
  // And in case of xyz
}
```

Also feels a bit weird, as the comment is not about the subsequent line, but about the whole `else`.

Hence I went for

```
if {
// Some case
}
else {
// And in case of xyz
}
```

for consistency.